rosetta/retriever/retriever_test.go

```diff
@@ -0,0 +1,1066 @@
+			TokensDepositedFunc: func(symbol string) (string, error) {
+				assert.Equal(t, symbol, dps.FlowSymbol)
+				return "", nil
+			},
+			TokensWithdrawnFunc: func(symbol string) (string, error) {
+				assert.Equal(t, symbol, dps.FlowSymbol)
+				return "", nil
+			},
+		}
+		index := &mocks.Reader{
+			HeaderFunc: func(height uint64) (*flow.Header, error) {
+				assert.Equal(t, testHeight, height)
+				return nil, mocks.DummyError
+			},
+		}
+
+		r := &Retriever{
+			validate:  validator,
+			generator: generator,
+			index:     index,
+		}
+
+		_, _, err := r.Block(testBlockID)
+		assert.Error(t, err)
+	})
+
+	t.Run("handles index retrieval failures", func(t *testing.T) {
+		validator := &mocks.Validator{
+			BlockFunc: func(block identifier.Block) (identifier.Block, error) {
+				assert.Equal(t, testBlockID, block)
+				return block, nil
+			},
+		}
+		generator := &mocks.Generator{
+			TokensDepositedFunc: func(symbol string) (string, error) {
+				assert.Equal(t, symbol, dps.FlowSymbol)
+				return "deposit", nil
+			},
+			TokensWithdrawnFunc: func(symbol string) (string, error) {
+				assert.Equal(t, symbol, dps.FlowSymbol)
+				return "withdrawal", nil
+			},
+		}
+		index := &mocks.Reader{
+			HeaderFunc: func(height uint64) (*flow.Header, error) {
+				assert.Equal(t, testHeight, height)
+				return testHeader, nil
+			},
+			EventsFunc: func(height uint64, types ...flow.EventType) ([]flow.Event, error) {
+				assert.Equal(t, testHeight, height)
+				if assert.Len(t, types, 2) {
+					assert.Equal(t, flow.EventType("deposit"), types[0])
+					assert.Equal(t, flow.EventType("withdrawal"), types[1])
+				}
+				return nil, mocks.DummyError
+			},
+		}
+
+		r := &Retriever{
+			validate:  validator,
+			generator: generator,
+			index:     index,
+		}
+
+		_, _, err := r.Block(testBlockID)
+		assert.Error(t, err)
+	})
+
+	t.Run("handles incorrectly-formatted indexed events", func(t *testing.T) {
+		validator := &mocks.Validator{
+			BlockFunc: func(block identifier.Block) (identifier.Block, error) {
+				assert.Equal(t, testBlockID, block)
+				return block, nil
+			},
+		}
+		generator := &mocks.Generator{
+			TokensDepositedFunc: func(symbol string) (string, error) {
+				assert.Equal(t, symbol, dps.FlowSymbol)
+				return "deposit", nil
+			},
+			TokensWithdrawnFunc: func(symbol string) (string, error) {
+				assert.Equal(t, symbol, dps.FlowSymbol)
+				return "withdrawal", nil
+			},
+		}
+		index := &mocks.Reader{
+			HeaderFunc: func(height uint64) (*flow.Header, error) {
+				assert.Equal(t, testHeight, height)
+				return testHeader, nil
+			},
+			EventsFunc: func(height uint64, types ...flow.EventType) ([]flow.Event, error) {
+				assert.Equal(t, testHeight, height)
+				if assert.Len(t, types, 2) {
+					assert.Equal(t, flow.EventType("deposit"), types[0])
+					assert.Equal(t, flow.EventType("withdrawal"), types[1])
+				}
+
+				return []flow.Event{{
+					Payload: []byte(`invalid_payload`),
+				}}, nil
+			},
```

Comment:
I would suggest to inject the event conversion function that I suggest further up, so that we can actually mock that out. That way, we don't have to cover the converter test cases, and can test conversion separately from the interaction of the components. So we would be able to remove this test case, and instead only have `handles events retrieval error` and `handle event conversion error`.